src/box/replication.cc

```diff
@@ -206,6 +209,70 @@ replicaset_set_sync_quorum(void)
 	replication_sync_quorum_auto = replicaset.applier.connected;
 }
 
+/**
+ * Deactivates all expired anonymous replicas. They are expected to be
+ * deleted on deactivation as orphan.
+ * Returns lowest time-to-live among remaining anonymous replicas via
+ * `lowest_ttl`. If there are no anonymous replicas left or the function has
+ * failed to deactivate a replica, `replication_anon_ttl` is returned.
+ */
+static void
+replication_anon_gc(double *lowest_ttl)
+{
+	while (true) {
+		*lowest_ttl = replication_anon_ttl;
+		struct replica *expired_replica = NULL;
+		/*
+		 * Search for the first expired replica. Note that we cannot
+		 * just deactivate it and continue the iteration because a yield
+		 * will happen and the replicaset can be changed significantly.
+		 */
+		replicaset_foreach(replica) {
+			if (!replica->anon || replica_has_connections(replica))
+				continue;
+			double last_seen = relay_last_row_time(replica->relay);
+			double unused = ev_monotonic_now(loop()) - last_seen;
+			double ttl = replication_anon_ttl - unused;
+			if (ttl <= 0) {
+				expired_replica = replica;
+				break;
+			}
+			*lowest_ttl = MIN(*lowest_ttl, ttl);
+		}
+		if (expired_replica == NULL)
+			break;
+		/* Save UUID since the replica will be deleted on success. */
+		struct tt_uuid uuid = expired_replica->uuid;
+		if (replica_deactivate(expired_replica) != 0) {
+			say_error("Failed to deactivate expired anonymous "
+				  "replica %s", tt_uuid_str(&uuid));
+			diag_log();
+			*lowest_ttl = replication_anon_ttl;
+			break;
+		}
+		/*
+		 * Disconnected anonymous replica without WAL GC state is
+		 * expected to be orphan, hence, deleted.
+		 */
+		assert(replica_by_uuid(&uuid) == NULL);
+		say_info("Anonymous replica %s has not been used for too long "
+			 "so it has been deactivated", tt_uuid_str(&uuid));
+	}
+}
+
+static int
+replication_anon_gc_fiber_f(va_list ap)
+{
+	(void)ap;
+	while (!fiber_is_cancelled()) {
+		fiber_check_gc();
+		double lowest_ttl = 0;
+		replication_anon_gc(&lowest_ttl);
```

Comment:
Since this function is called in a single place and already has a loop in it, I think it'd be better to inline it here, replacing `break` with `sleep` + `continue`.